### PR TITLE
[Xamarin.Android.Build.Utilities] Add AndroidVersions

### PR DESCRIPTION
**Pull request by jonpryor**
Commit 8e7d37bb is a sign of duplication: in order to use a
`Mono.Android.dll` binding assembly, not only do we need to *build*
the binding assembly, but we *also* need to update
`Xamarin.Android.Build.Utilities.dll` to "know" about the new binding
version (to map API level to `$(TargetFrameworkVersion)`).

Even "better" (worse), if the new API level is a *preview*, there is
no *consistent* API level. For example, with API-O, `android.jar`
isn't at
`$(AndroidSdkDirectory)\platforms\android-@API_LEVEL@\android.jar`,
where `@API_LEVEL@` is 26 (because various codepaths require that the
"api level" be an integer). Instead, it's installed at
`$(AndroidSdkDirectory)\platforms\android-O\android.jar`, where
`O` is the "id" of the preview API level.

This "id" is "leaky", in turn requiring that
`Xamarin.Android.Build.Tasks.dll` *also* be updated to deal with the
mappings.

Even "better" (worse), if we *forget* to cross all our our 't's and
dot all of our 'i's, we'll have a binding assembly which can't be
used. (Which is why we needed commit 8e7d37bb; without it, the API-O
binding can't be used!)

This is all obviously madness. ;-)

Clean this mess up:

 1. Update `src/Mono.Android` to create a new `AndroidApiInfo.xml`
    file within the `$(TargetFrameworkDriectory)` directory.
    This will contain all the information needed to map Android API
    levels to Ids and Android OS versions and
    `$(TargetFrameworkVersion)` values.

    ```
    <AndroidApiInfo>
      <Id>10</Id>
      <Level>10</Level>
      <Name>Gingerbread</Name>
      <Version>v2.3</Version>
    </AndroidApiInfo>
    ```

 2. Add a new `Xamarin.Android.Build.Utilities.AndroidVersions` type
    which looks for and parses these new `AndroidApiInfo.xml` files.

 3. Fixup all the other places using `AndroidVersion.KnownVersions`
    and related members to instead use `AndroidVersions`.

 4. Deprecate all the old APIs which rely on hardcoded data.

The advantage to all this is that we can support new API level
bindings by just building a new `Mono.Android.dll` and placing an
`AndroidApiInfo.xml` into the appropriate location (next to
`Mono.Android.dll`). No further code changes would be required.

### CODE REVIEW

File: src/Xamarin.Android.Build.Utilities/AndroidVersion.cs [hunk 1]
```diff
@@ -4,6 +4,7 @@ namespace Xamarin.Android.Build.Utilities
 {
 	public class AndroidVersion
 	{
+		[Obsolete ("Use AndroidVersions.MaxVersion.ApiLevel", error:true)]
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Utilities/AndroidVersion.cs]: ...which doesn't exist. Oops.

Anyway, it's not used in this project, due to `error:true`. Do we really need to track `MaxApiLevel`?

**Comment by garuma** [comment]
Excellent. Now the question is how do we propagate that to the IDEs. I see two options:

- Update androidtools to also lookup that file instead of the hardcoded list. It makes it possible to have newer XA work with the same IDE but we still get code duplication (albeit we don't have to make change to it)
- Expose that API information through an MSBuild target outputting items with the right metadata that is stored by the IDE. No androidtools needed.

Both approach have pro and cons, first might be easier to implement for now though

**Comment by jonpryor** [comment]
> * Expose that API information through an MSBuild target outputting items with the right metadata that is stored by the IDE. No androidtools needed.

That's an interesting idea, but if we went for the MSBuild-specific approach, why bother with half of this code at all? We could provide `AndroidApiInfo.targets` files, one per API level, and `<Import/>` the mess of them within a `<Target/>`.

(That might not be a bad idea, actually. More thought is required.)

As for the integration idea, there is already (kinda, sorta) a plan (-ish) for that: [xamarin-android-tools](https://github.com/xamarin/xamarin-android-tools), which is kinda/sorta the "open-source" parts of `androidtools`. We'd "just" need to update the IDEs to use `xamarin-android-tools`.

*However*, I want to do some API cleanups ("breaks"!) as part of this, which is a discussion for another time. It's basically the *only* time we can do such cleanups, as it would be in a new assembly...

**Comment by mhutch** [comment]
I don't think it really make sense to get this info though MSBuild calls since it doesn't depend on the state of the project at all.

**Comment by jonpryor** [comment]
The problem in [**macOS+xbuild PR Build 952**](https://jenkins.mono-project.com/job/xamarin-android-pr-builder/952/) is our old nemesis, xbuild, which [looks at `GetReferenceAssemblyPaths.RootPath` *after* `$XBUILD_FRAMEWORK_FOLDERS_PATH` and `/Library/Frameworks/Mono.framework/External`](https://github.com/mono/mono/blob/mono-4.8.0-branch/mcs/class/Microsoft.Build.Tasks/Microsoft.Build.Tasks/GetReferenceAssemblyPaths.cs#L86-L90). This means that when `Xamarin.Android.NUniteLite.csproj` is built, the wrong directory used:

```
Building target "_GetReferenceAssemblyPaths" in project "/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Xamarin.Android.NUnitLite/Xamarin.Android.NUnitLite.csproj" ("/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/bin/Debug/lib/xbuild/Xamarin/Android/Xamarin.Android.Common.targets"); "_SetLatestTargetFrameworkVersion" depends on it.
         Target _GetReferenceAssemblyPaths:
         Task "GetReferenceAssemblyPaths"
                 Using task GetReferenceAssemblyPaths from Microsoft.Build.Tasks.GetReferenceAssemblyPaths, Microsoft.Build.Tasks.Core, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
                 Looking for framework 'MonoAndroid,Version=v1.0' in root path '/Library/Frameworks/Mono.framework/External/xbuild-frameworks'
                 Found framework definition list '/Library/Frameworks/Mono.framework/External/xbuild-frameworks/MonoAndroid/v1.0/RedistList/FrameworkList.xml' for framework 'MonoAndroid,Version=v1.0'
         Done executing task "GetReferenceAssemblyPaths"
         Done building target "_GetReferenceAssemblyPaths" in project "/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Xamarin.Android.NUnitLite/Xamarin.Android.NUnitLite.csproj".
Done building target "_GetReferenceAssemblyPaths" in project "/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Xamarin.Android.NUnitLite/Xamarin.Android.NUnitLite.csproj" ("/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/bin/Debug/lib/xbuild/Xamarin/Android/Xamarin.Android.Common.targets"); "_SetLatestTargetFrameworkVersion" depends on it.
```

Since the *system-wide* directory is used here -- a directory which does not, *cannot*, contain the new `AndroidApiInfo.xml` files -- everything that depends on `AndroidVersions` fails. (This despite the fact that I *do not want* to be using the system-wide directory!)

There are two plausible fixes that bear investigation:

1. Dump `xbuild`, migrate to `msbuild`. Note that the [**macOS+msbuild PR Build**](https://jenkins.mono-project.com/job/xamarin-android-msbuild-pr-builder/467/) *succeeded*. This is because msbuild is awesome, except for the terrible parts I hate.

    The problem with this is that msbuild on Linux is a bit of an unknown; @atsushieno has lost lots of time trying to get msbuild working on his machine. If there's a Linux distro with a prebuilt `msbuild` which *works*, that's one thing, but that's not the situation for everyone.

2. Don't build with `xbuild`, build with `xabuild` (which overrides `$XBUILD_FRAMEWORK_FOLDERS_PATH`).

     This *should* allow things to work, but I'm not sure what the downlevel effects may be.

**Comment by jonpryor** [comment]
Next in our comedy of "ow! make it stop!", using `xabuild` to build things *kind of* worked: [**macOS+xbuild** was able to build](https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android-pr-builder/1250/) (if not pass all the unit tests), while [**macOS+msbuild** couldn't build](https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android-msbuild-pr-builder/773/), because `msbuild` couldn't resolve `.NETFramework,Version=v4.6.1`.

Fix *that* issue by improving `xabuild` to not only symlink the contents of `$mono_prefix/lib/mono/xbuild/*` into `bin/Debug/lib/xamarin.android/xbuild`, but to *also* symlink the contents of `$mono_prefix/lib/mono/xbuild-frameworks/*` into `bin/Debug/lib/xamarin.android/xbuild-frameworks`. [This allows **macOS+msbuild** to build](https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android-msbuild-pr-builder/782/), but [broke **macOS+xbuild**](https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android-pr-builder/1259/).

(Back and forth we swing!)

The most recent failure is "fun": [the `<GetReferenceAssemblyPaths/>` task reads `document('FrameworkList.xml')/FileList/@TargetFrameworkDirectory`, and appends that to the value of `$(TargetFrameworkRootPath)`/the probed framework directory](https://github.com/mono/mono/blob/mono-4.8.0-branch/mcs/class/Microsoft.Build.Tasks/Microsoft.Build.Tasks/GetReferenceAssemblyPaths.cs#L138-L142).

For context, `redistlist_dir=bin/Debug/lib/xamarin.android/xbuild-frameworks/.NETFramework/v4.6.1/RedistList`, and `framework_dir=..\..\..\..\4.6.1-api`. The result is that, after `<GetReferenceAssemblyPaths/>` completes, `xbuild` is trying to look for assemblies in the path `bin/Debug/lib/xamarin.android/4.6.1-api`, which *does not exist*.

(Hacking `xbuild` internals for fun and profit?)

At this point, I'm not sure how to fix this, much less *elegantly* fix this, short of "kill xbuild, msbuild all the things!"

Which, to be fair, is looking *really* appealing right now...

File: src/Mono.Android/Mono.Android.projitems [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      <Name>Nougat</Name>
+      <Level>25</Level>
+      <Id>25</Id>
+      <Stable>True</Stable>
+    </AndroidApiInfo>
+    <AndroidApiInfo Include="v8.0">
+      <Name>O Preview</Name>
+      <Level>26</Level>
+      <Id>O</Id>
+      <Stable>False</Stable>
```
Comment (jonpryor) [inline src/Mono.Android/Mono.Android.projitems]: Should probably change this to True...

File: src/Xamarin.Android.Build.Utilities/AndroidAppManifest.cs [hunk 3]
```diff
@@ -166,7 +166,7 @@ static string NullIfEmpty (string value)
 				return null;
 			int vn;
 			if (!int.TryParse (version, out vn))
-				vn = AndroidVersion.MaxApiLevel;
+				return null;
```
Comment (dellis1972) [inline src/Xamarin.Android.Build.Utilities/AndroidAppManifest.cs]: @jonpryor should we be returning null here?

File: src/Mono.Android/Mono.Android.targets [hunk 4]
```diff
@@ -120,6 +121,31 @@
...
+     />
+     <_Line Include="&lt;AndroidApiInfo&gt;" />
+     <_Line Include="@(_ApiInfo->'  &lt;Id>%(Id)&lt;/Id>')" />
+     <_Line Include="@(_ApiInfo->'  &lt;Level>%(Level)&lt;/Level>')" />
+     <_Line Include="@(_ApiInfo->'  &lt;Name>%(Name)&lt;/Name>')" />
+     <_Line Include="@(_ApiInfo->'  &lt;Version>%(Identity)&lt;/Version>')" />
+     <_Line Include="@(_ApiInfo->'  &lt;Stable>%(Stable)&lt;/Stable>')" />
+     <_Line Include="&lt;/AndroidApiInfo&gt;" />
+   </ItemGroup>
+   <Message Text="# jonp: _Line=@(_Line)" />
```
Comment (dellis1972) [inline src/Mono.Android/Mono.Android.targets]: @jonpryor can we remove the debug message :)

File: src/Xamarin.Android.Build.Utilities/AndroidAppManifest.cs (hunk 3 above)
Comment (dellis1972) [inline src/Xamarin.Android.Build.Utilities/AndroidAppManifest.cs]: I guess this should be `AndroidVersions. MaxStableVersion` ?

**Pull request merged**